File: src/main/java/com/severett/chargerapp/data/InMemorySessionStatisticsRepo.java
package com.severett.chargerapp.data;

import org.springframework.stereotype.Repository;

import java.time.Instant;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Repository
public class InMemorySessionStatisticsRepo implements SessionStatisticsRepo {

    private final Map<Long, Long> startSessionCounter;

    private final Map<Long, Long> stopSessionCounter;

    public InMemorySessionStatisticsRepo() {
        startSessionCounter = new ConcurrentHashMap<>();
        stopSessionCounter = new ConcurrentHashMap<>();
    }

    @Override
    public void addSessionStart(Instant timestamp) {
        startSessionCounter.compute(
                timestamp.getEpochSecond(),
                (key, oldValue) -> oldValue != null ? oldValue + 1L : 1L
        );
    }

    @Override
    public void addSessionStop(Instant timestamp) {
        stopSessionCounter.compute(
                timestamp.getEpochSecond(),
                (key, oldValue) -> oldValue != null ? oldValue + 1L : 1L
        );
    }

    @Override
    public long getSessionStartCount(Instant fromTimestamp, Instant toTimestamp) {
        return getCountAggregate(
                startSessionCounter,
                fromTimestamp.getEpochSecond(),
                toTimestamp.getEpochSecond()
        );
    }

    @Override
    public long getSessionStopCount(Instant fromTimestamp, Instant toTimestamp) {
        return getCountAggregate(
                stopSessionCounter,
                fromTimestamp.getEpochSecond(),
                toTimestamp.getEpochSecond()
        );
    }

    private long getCountAggregate(Map<Long, Long> counter, long fromEpoch, long toEpoch) {
        long retValue = 0L;
        long currentEpoch = fromEpoch;
        do {
            retValue += counter.getOrDefault(currentEpoch, 0L);
            currentEpoch++;
        } while (currentEpoch <= toEpoch);
        return retValue;
    }
}

File: src/main/kotlin/com/severett/chargerapp/data/InMemorySessionStatisticsRepo.kt
package com.severett.chargerapp.data

import org.springframework.stereotype.Repository
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

@Repository
class InMemorySessionStatisticsRepo : SessionStatisticsRepo {

    private val startSessionCounter = ConcurrentHashMap<Long, Long>()

    private val stopSessionCounter = ConcurrentHashMap<Long, Long>()

    override fun addSessionStart(timestamp: Instant) {
        startSessionCounter.compute(timestamp.epochSecond) { _, oldValue ->
            oldValue?.let { it + 1L } ?: 1L
        }
    }

    override fun addSessionStop(timestamp: Instant) {
        stopSessionCounter.compute(timestamp.epochSecond) { _, oldValue ->
            oldValue?.let { it + 1L } ?: 1L
        }
    }

    override fun getSessionStartCount(fromTimestamp: Instant, toTimestamp: Instant): Long =
        getCountAggregate(startSessionCounter, fromTimestamp.epochSecond, toTimestamp.epochSecond)

    override fun getSessionStopCount(fromTimestamp: Instant, toTimestamp: Instant): Long =
        getCountAggregate(stopSessionCounter, fromTimestamp.epochSecond, toTimestamp.epochSecond)

    fun getCountAggregate(counter: Map<Long, Long>, fromEpoch: Long, toEpoch: Long): Long {
        return (fromEpoch..toEpoch).fold(0L) { retVal, currentEpoch ->
            retVal + counter.getOrDefault(currentEpoch, 0L)
        }
    }

}
